File: Structural/Flyweight/python/FlyWheight.py
class FlyWheight():

    def __init__(self, shared_state):
        self.shared_state = shared_state


class Context():

    def __init__(self, unique_state, flywheight: FlyWheight):
        self.unique_state = unique_state
        self.flywheight = flywheight

    def __repr__(self):
        return "unique: %s - shared: %s" % (self.unique_state, self.flywheight.shared_state)

class FlyWheightFactory():

    def __init__(self):
        self.flywheights = []

    def get_flyweight(self, shared_state) -> FlyWheight:

        flywheights = list(filter(lambda x: x.shared_state ==
                                  shared_state, self.flywheights))
        if len(flywheights) > 0:
            return flywheights[0]
        else:
            flywheight = FlyWheight(shared_state)
            self.flywheights.append(flywheight)
            return flywheight

    @property
    def total(self):
        return len(self.flywheights)


class Client():

    def __init__(self, flywheight_factory: FlyWheightFactory):
        self.flywheight_factory = flywheight_factory
        self.contexts = []

    def get_context(self, unique_state, shared_state) -> Context:
        
        flywheight = self.flywheight_factory.get_flyweight(shared_state)
        context = Context(unique_state, flywheight)
        self.contexts.append(context)
        
        return context


def main():
    flywheight_factory = FlyWheightFactory()
    client = Client(flywheight_factory)

    shared_states = ['Alpha', 'Beta', 'Gamma']
    unique_states = list(range(10))

    contexts = [client.get_context(x, y) for x in unique_states for y in shared_states]

    print("Contexts:", len(contexts))
    print("Flywheights:", flywheight_factory.total)
    

if __name__ == "__main__":
    main() 

File: Structural/Flyweight/scala/Flywheight.scala
import scala.collection.mutable.ListBuffer

case class Flyweight(sharedState: String)

case class Context(uniqueState: Int, flyweight: Flyweight)

class FlyweightFactory {

  var flyweights = ListBuffer.empty[Flyweight]

  def getFlyweight(sharedState: String): Flyweight = {
    val selectedFlyweights = flyweights.filter(flyweight => flyweight.sharedState.equals(sharedState))

    if (selectedFlyweights.size > 0) {
      selectedFlyweights(0)
    } else {
      val newFlyweight = Flyweight(sharedState)
      flyweights.addOne(newFlyweight)
      newFlyweight
    }
  }

  def total = flyweights.size
}


class Client(flyweightFactory: FlyweightFactory) {

  val contexts = ListBuffer.empty[Context]

  def getContext(uniqueState: Int, sharedState: String): Context = {
    val flyweight = flyweightFactory.getFlyweight(sharedState)
    val context = Context(uniqueState, flyweight)
    contexts.addOne(context)

    context
  }

}

object FlyweightPattern {

  def main(args: Array[String]): Unit = {
    val flyweightFactory = new FlyweightFactory()
    val client = new Client(flyweightFactory)


    val uniqueStates = Range(0, 10).toSeq
    val sharedStates = Seq("Alpha", "Beta", "Gamma")

    val contexts = uniqueStates.flatMap(uniqueState => {
      sharedStates.map(sharedState => client.getContext(uniqueState, sharedState))
    })

    println(f"Contexts: ${contexts.size}")
    println(f"Flyweights: ${flyweightFactory.total}")
  }
}
